### PR TITLE
Add movie and category, user can add movies, can add categories, can …

### PR DESCRIPTION
**Pull request by bodikqlar**
…add categories to movies.

### CODE REVIEW

File: app/controllers/categories_controller.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    if @category.save
+      render "index"
+    else
+      render "new"
+    end
+  end
+
+  def index
+    @categories = Category.all
+    p params
```
Comment (maksymst) [inline app/controllers/categories_controller.rb]: remove

File: app/controllers/categories_controller.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    respond_to do |format|
+      format.html
+      format.js
```
Comment (MaximSt) [inline app/controllers/categories_controller.rb]: why do you need this?

File: app/controllers/movies_controller.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+class MoviesController < ApplicationController
+  def new
+    @movie = Movie.new
+  end
+
+  def create
+    movie_attrs = movie_params
+    movie_attrs[:user] = current_user
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: use associations here?

File: app/controllers/movies_controller.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
+class MoviesController < ApplicationController
+  def new
+    @movie = Movie.new
+  end
+
+  def create
+    movie_attrs = movie_params
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: redundant

File: app/controllers/movies_controller.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  def index
+    @movies = current_user.movies
+  end
+
+  def destroy
+    category = Movie.find(params[:id]).destroy
+    redirect_to action: "index"
+  end
+
+  def assign_category
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: read more about `has_and_belongs_to_many`

File: app/controllers/movies_controller.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    movie = user.movies.find(params[:id])
+    category = Category.find(params[:category_id])
+    movie.categories << category if !movie.categories.exists?(category)
+  end
+
+  private
+  def movie_params
+    p params.require(:movie).permit(:name, :rating)
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: `p` is redundant

File: app/models/user.rb [hunk 7]
```diff
@@ -1,6 +1,8 @@
 class User < ActiveRecord::Base
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable and :omniauthable
+  has_many :movies
```
Comment (maksymst) [inline app/models/user.rb]: where is `dependent` option?

File: app/views/categories/index.html.haml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+%h1Categories
+- if @categories
```
Comment (maksymst) [inline app/views/categories/index.html.haml]: what the purpose of `if`

File: app/views/categories/index.html.haml [hunk 9]
```diff
@@ -0,0 +1,4 @@
+%h1Categories
```
Comment (maksymst) [inline app/views/categories/index.html.haml]: space is missing

File: app/views/categories/index.js.erb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+<%- @categories.each do |category|%>
+        $(".categories").append("<div><div><%=category.name%></div> <%= link_to 'Add category', assign_category_movie_path(params[:movie_id], category_id: category.id), method: :put, remote: true%></div>");
+
```
Comment (maksymst) [inline app/views/categories/index.js.erb]: extra lines

File: app/views/categories/index.js.erb [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+
+<%end %>
```
Comment (maksymst) [inline app/views/categories/index.js.erb]: missing space

File: app/views/movies/index.html.haml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+%h1Categories
+- if @movies
```
Comment (maksymst) [inline app/views/movies/index.html.haml]: what the purpose of `if`

File: app/views/movies/index.html.haml [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+  - @movies.each do |movie|
+    .movie
```
Comment (maksymst) [inline app/views/movies/index.html.haml]: why not just use table?

File: app/views/movies/index.html.haml [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+      %span Name
+      %span.movie_name= movie.name
+      %span Rating
+      %span.movie_rating= movie.rating
+      - if !movie.categories.empty?
```
Comment (maksymst) [inline app/views/movies/index.html.haml]: remove this

File: db/migrate/20150506141629_create_movies.rb [hunk 15]
```diff
@@ -0,0 +1,10 @@
+class CreateMovies < ActiveRecord::Migration
+  def change
+    create_table :movies do |t|
+      t.string :name
+      t.integer :rating
+      t.belongs_to :user
```
Comment (maksymst) [inline db/migrate/20150506141629_create_movies.rb]: where is index?

File: db/migrate/20150506142018_create_categories_movies_join_table.rb [hunk 16]
```diff
@@ -0,0 +1,8 @@
+class CreateCategoriesMoviesJoinTable < ActiveRecord::Migration
+  def change
+    create_table :categories_movies, id: false do |t|
+      t.integer :category_id
```
Comment (maksymst) [inline db/migrate/20150506142018_create_categories_movies_join_table.rb]: use `belongs_to`

File: spec/models/category_spec.rb [hunk 17]
```diff
@@ -0,0 +1,14 @@
+require 'rails_helper'
+
+RSpec.describe Category, type: :model do
+  it { should validate_presence_of(:name) }
```
Comment (maksymst) [inline spec/models/category_spec.rb]: use `is_expected`

File: spec/models/category_spec.rb [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+  it { should have_and_belong_to_many(:movies) }
+
+  describe "Category" do
+    let(:category) { FactoryGirl.create(:category) }
```
Comment (maksymst) [inline spec/models/category_spec.rb]: include `FactoryGirl` into rails_helper

File: spec/models/movie_spec.rb [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+RSpec.describe Movie, type: :model do
+  it { should validate_presence_of(:name) }
+  it { should have_and_belong_to_many(:categories) }
+  it { should belong_to(:user) }
+
+  describe "Movie" do
+    let(:user) { FactoryGirl.create(:user) }
+    let(:movie) { FactoryGirl.create(:movie, user: user) }
+
+    it "check the movie" do
```
Comment (maksymst) [inline spec/models/movie_spec.rb]: why do you need this test?

File: app/controllers/movies_controller.rb [hunk 20]
```diff
@@ -0,0 +1,48 @@
+class MoviesController < ApplicationController
+  before_action :movie, only: [:destroy, :update, :edit]
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: use %i(...) for symbols array with length > 1

File: app/controllers/movies_controller.rb [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+      render "new"
+    end
+  end
+
+  def index
+    @movies = current_user.movies
+  end
+
+  def destroy
+    movie.destroy
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: why do you use method `movie` here?

File: app/views/movies/index.html.haml [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+    %tr
+      %th Name
+      %th Rating
+      %th Categories
+  - @movies.each do |movie|
+    %tbody
+      %tr
+        %td= movie.name
+        %td= movie.rating
+        %td= movie.categories.map(&:name).join(', ')
```
Comment (maksymst) [inline app/views/movies/index.html.haml]: you can use `pluck` instead of `map` for this purpose

File: app/views/movies/index.html.haml [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+        %td= link_to 'Edit', edit_movie_path(id: movie.id)
```
Comment (maksymst) [inline app/views/movies/index.html.haml]: you can write `edit_movie_path(movie.id)` instead

File: app/views/movies/index.html.haml [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+        %td= link_to 'Delete', movie_path(id: movie.id) , method: :delete
```
Comment (maksymst) [inline app/views/movies/index.html.haml]: you can write `movie_path(movie.id)` instead

File: spec/models/category_spec.rb [hunk 25]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
+RSpec.describe Category, type: :model do
+  it { is_expected.to  validate_presence_of(:name) }
```
Comment (maksymst) [inline spec/models/category_spec.rb]: extra space after `is_expected.to`

File: spec/models/category_spec.rb [hunk 26]
```diff
@@ -0,0 +1,10 @@
...
+  it { is_expected.to  have_and_belong_to_many(:movies) }
```
Comment (maksymst) [inline spec/models/category_spec.rb]: extra space after `is_expected.to`

File: spec/models/movie_spec.rb [hunk 27]
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
+RSpec.describe Movie, type: :model do
+  it { is_expected.to  validate_presence_of(:name) }
```
Comment (maksymst) [inline spec/models/movie_spec.rb]: extra space after `is_expected.to`

File: spec/models/user_spec.rb [hunk 28]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  it { is_expected.to  validate_presence_of(:email) }
```
Comment (maksymst) [inline spec/models/user_spec.rb]: redundant test

File: app/views/categories/index.html.haml [hunk 29]
```diff
@@ -0,0 +1,5 @@
+%h1 Categories
+- @categories.each do |category|
+    .category= category.name
```
Comment (maksymst) [inline app/views/categories/index.html.haml]: fix indent

File: app/views/movies/_form.html.haml [hunk 30]
```diff
@@ -0,0 +1,10 @@
+= form_for(@movie) do |f|
+  .name
+    = f.label :name
+    = f.text_field :name
+  .thumburl
+    = f.label :rating
+    = f.text_field :rating
+    = f.collection_select :categories, Category.all, :id, :name
```
Comment (maksymst) [inline app/views/movies/_form.html.haml]: do not use logic on view

File: app/views/movies/new.html.haml [hunk 31]
```diff
@@ -0,0 +1,2 @@
+%h1 Create new movie
+=render partial: "form"
```
Comment (maksymst) [inline app/views/movies/new.html.haml]: missing space

File: app/controllers/movies_controller.rb [hunk 32]
```diff
@@ -0,0 +1,52 @@
...
+    @movie.destroy
+    redirect_to movies_path
+  end
+
+  def edit
+  end
+
+  def update
+    if category_ids = params[:movie][:categories]
+      movie.category_ids= category_ids
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: missing space

File: app/views/movies/index.html.haml [hunk 33]
```diff
@@ -0,0 +1,18 @@
+%h1 Movies
+%table
+  %thead
+    %tr
+      %th Name
+      %th Rating
+      %th Categories
+  %tbody
+  - @movies.each do |movie|
```
Comment (maksymst) [inline app/views/movies/index.html.haml]: add indent

File: app/helpers/application_helper.rb [hunk 34]
```diff
@@ -13,7 +13,17 @@ def modal_helper(header, modal, &block)
 end
 
 def show_header(header)
```
Comment (maksymst) [inline app/helpers/application_helper.rb]: why all methods below are outside of ApplicationHelper?

File: app/controllers/categories_controller.rb [hunk 35]
```diff
@@ -0,0 +1,46 @@
...
+  before_action :category, only: %i(destroy update edit)
+
+  def new
+    @category = Category.new
+  end
+
+  def create
+    @category = Category.new(category_params)
+    if @category.save
+      redirect_to categories_path
```
Comment (maksymst) [inline app/controllers/categories_controller.rb]: add localized flash messages for all actions

File: app/controllers/movies_controller.rb [hunk 36]
```diff
@@ -0,0 +1,52 @@
...
+    @movie = current_user.movies.build(movie_params)
+    if @movie.save
+      redirect_to movies_path
+    else
+      render :new
+    end
+  end
+
+  def index
+    @movies = Movie.includes(:user).all
```
Comment (maksymst) [inline app/controllers/movies_controller.rb]: `all` is redundant

File: spec/controllers/categories_controller_spec.rb [hunk 37]
```diff
@@ -0,0 +1,122 @@
...
+    end
+  end
+
+  describe 'PUT #update' do
+    context 'with valid params' do
+      let(:new_attributes) { { name: 'Yuhu'} }
+      before(:each) { put :update, {id: category.id, category: new_attributes} }
+
+      it 'updates the requested category' do
+        category.reload
```
Comment (maksymst) [inline spec/controllers/categories_controller_spec.rb:107]: test is missing

File: spec/controllers/categories_controller_spec.rb [hunk 38]
```diff
@@ -0,0 +1,122 @@
...
+      end
+
+      it 'redirects to the categories' do
+        expect(response).to redirect_to categories_path
+      end
+    end
+
+    context 'with invalid params' do
+      let(:unvalid_attributes) {{ name: nil }}
```
Comment (maksymst) [inline spec/controllers/categories_controller_spec.rb]: `invalid_attributes`

File: spec/controllers/movies_controller_spec.rb [hunk 39]
```diff
@@ -0,0 +1,124 @@
...
+    end
+  end
+
+  describe 'PUT #update' do
+    context 'with valid params' do
+      let(:new_attributes) { { name: 'Yuhu', rating: 2} }
+      before(:each) { put :update, {id: movie.id, movie: new_attributes} }
+
+      it 'updates the requested movie' do
+        movie.reload
```
Comment (maksymst) [inline spec/controllers/movies_controller_spec.rb:109]: this test does nothing

**Pull request merged**